pkg/operator/controllers/machine/machine.go

```diff
@@ -38,16 +39,27 @@ func (r *Reconciler) machineValid(ctx context.Context, machine *machinev1beta1.M
 		return []error{fmt.Errorf("machine %s: provider spec missing", machine.Name)}
 	}
 
-	o, _, err := scheme.Codecs.UniversalDeserializer().Decode(machine.Spec.ProviderSpec.Value.Raw, nil, nil)
-	if err != nil {
-		return []error{err}
-	}
+	var machineProviderSpec *machinev1beta1.AzureMachineProviderSpec
 
-	machineProviderSpec, ok := o.(*azureproviderv1beta1.AzureMachineProviderSpec)
-	if !ok {
-		// This should never happen: codecs uses scheme that has only one registered type
-		// and if something is wrong with the provider spec - decoding should fail
-		return []error{fmt.Errorf("machine %s: failed to read provider spec: %T", machine.Name, o)}
+	if strings.Contains(string(machine.Spec.ProviderSpec.Value.Raw), "azureproviderconfig.openshift.io") {
+		machineProviderSpec = &machinev1beta1.AzureMachineProviderSpec{}
+		err := json.Unmarshal(machine.Spec.ProviderSpec.Value.Raw, machineProviderSpec)
+		if err != nil {
+			return []error{fmt.Errorf("machine %s: failed to unmarshal the 'azureproviderconfig.openshift.io' provider spec: %q", machine.Name, err.Error())}
+		}
+	} else {
+		o, _, err := scheme.Codecs.UniversalDeserializer().Decode(machine.Spec.ProviderSpec.Value.Raw, nil, nil)
+		if err != nil {
+			return []error{err}
+		}
+
+		var ok bool
+		machineProviderSpec, ok = o.(*machinev1beta1.AzureMachineProviderSpec)
+		if !ok {
+			// This should never happen: codecs uses scheme that has only one registered type
+			// and if something is wrong with the provider spec - decoding should fail
+			return []error{fmt.Errorf("machine %s: failed to read provider spec: %T", machine.Name, o)}
+		}
```

Comment:
Why do we need both "unmarshals" here?  Is it to support the older types as well as the newer types?

Reply:
For previous versions on OCP the **AzureMachineProviderSpec** is defined with `azureproviderconfig.openshift.io/v1beta1` but that has been moved to `machine.openshift.io/v1beta1` in 4.10:
https://github.com/openshift/installer/commit/f9725ddd9496b999ac06bd0343791d7c50b00d48#diff-54168990f7f297f463e06a145f714bd63582499b41b4be695f49f5f16dc6f76eR110

In this release we are moving to the new API, but we also need to be able to verify machines in all of the existing machinesets in the fleet, which are defined under `azureproviderconfig.openshift.io`. To simplify the switchover the raw string is checked for the old format and that is marshalled into the "new" `machineapi.AzureMachineProviderSpec` struct, which has exactly the same definition.

Note that the `machineapi.AzureMachineProviderSpec` is also not registered by default, so there is also this related change to ensure the operator machine validation still works:
https://github.com/darthhexx/ARO-RP/blob/2ba04295d89f14472c909530c262fda3e088d91e/pkg/util/scheme/scheme.go#L43-L44